src/main/java/bio/terra/common/retry/SamRetry.java

```diff
@@ -0,0 +1,136 @@
+package bio.terra.common.retry;
+
+import static java.time.Instant.now;
+
+import bio.terra.common.exception.InternalServerErrorException;
+import bio.terra.common.exception.SamApiException;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.google.common.annotations.VisibleForTesting;
+import java.time.Duration;
+import java.time.Instant;
+import java.util.concurrent.TimeUnit;
+import org.broadinstitute.dsde.workbench.client.sam.ApiException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.http.HttpStatus;
+
+/**
+ * SamRetry encapsulates logic needed for retrying Sam API calls. All constants are hard-coded, as
+ * Sam failures are generally just short periods of down-time.
+ * SamRetry always throws a SamApiException, which extracts the error message from Sam (which can be
+ * found in two different places) and stores the HTTP Status from Sam.
+ */
+public class SamRetry {
+  private static Logger logger = LoggerFactory.getLogger(SamRetry.class);
+
+  // The retry function starts with INITIAL_WAIT between retries, and doubles that until it
+  // reaches MAXIMUM_WAIT, after which all retries are MAXIMUM_WAIT apart.
+  private static final Duration MAXIMUM_WAIT = Duration.ofSeconds(30);
+  private static final Duration INITIAL_WAIT = Duration.ofSeconds(10);
+  private static final Duration OPERATION_TIMEOUT = Duration.ofSeconds(300);
+  private final Instant operationTimeout;
+
+  // How long to wait between retries.
+  private Duration retryDuration;
+
+  private static final ObjectMapper objectMapper = new ObjectMapper();
+
+  SamRetry() {
+    this.operationTimeout = now().plus(OPERATION_TIMEOUT);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  protected SamRetry(Duration timeout) {
+    this.operationTimeout = now().plus(timeout);
+    this.retryDuration = INITIAL_WAIT;
+  }
+
+  @FunctionalInterface
+  public interface SamVoidFunction {
+    void apply() throws ApiException, InterruptedException;
```

Comment:
The important thing is not to catch InterruptedExceptions while waiting. Those exceptions happen when Stairway needs to shut down its thread pool or otherwise cancel a flight.

If they never arise from Sam, then it isn't needed.

Reply:
How would we verify that Sam doesn't throw these? Also is it possible that SamService in WSM doesn't, but somewhere else it might?

Reply:
Since InterruptedException is a checked exception, I don't think the code compiles if the underlying modules throw those exceptions.

Reply:
Swallowing InterruptedException is a BadThing™
It stops us from gracefully shutting down flights.

Reply:
Good to know, I think we're doing that incorrectly in buffer in a couple places.

Reply:
Thank @wchamber for pinging me this. Good to know. Now InterruptedException is heavily used in Buffer as it need to keep polling some GCP operations result. I will change buffer's this behavior

Reply:
If stairway stops a flight, would the interrupted exception thrown from the sleep call really have information useful to stairway? That it expects to get back? How would that work?

Reply:
The interrupted exception is generated when Stairway asks its thread pool to terminate. The thread pool generates interrupts on all threads. The top level of the flight controller catches the InterruptedException. It disowns the flight, and if the queue is enabled, it queues the flight. It then returns, which returns the thread back to the thread pool. It doesn't need any extra information to do this.

Not all flights will see an InterruptedException. Only ones in wait states. Other flights may notice the request to shutdown at a step boundary where they check a shutting-down state flag and perform the same set of exit state changes.

The pool terminate times out at some point, and remaining flight threads will just be killed as the pod is shutdown. Those will have to be recovered by other pods or the restart of this pod.

Reply:
Okey dokey, InterruptedExceptions are being raised!